### DISCUSSION TITLE
Off Topic: Your NuGet EventWatcher Package

### DISCUSSION THREAD

**Discussion by BourgeoisDirk** [dotnettools/SharpGrabber#85] [General]
Dear Dev,

I'm having some trouble with one of your Packages, and I don't know why it occurs.
Thought I'd come by and ask ?

I've added "EventHookCore" to my Class Library, and added this class (at first i spoke to ClipboardWatcher directly, but changed to this)
`    using EventHook;

    public class EventWatcher
    {
        public delegate void DelegateClipboardChanged(ClipboardContentTypes format, object data);
        public event DelegateClipboardChanged ClipboardChangedEvent;
        internal void ClipboardChanged(ClipboardContentTypes format, object data)
            => ClipboardChangedEvent?.Invoke(format, data);

        private EventHookFactory EventFactory;
        private ClipboardWatcher ClipboardWatcher;

        public EventWatcher()
        {
            EventFactory = new EventHookFactory();
            ClipboardWatcher = EventFactory.GetClipboardWatcher();
            ClipboardWatcher.OnClipboardModified += ClipboardWatcher_OnClipboardModified;
        }

        private void ClipboardWatcher_OnClipboardModified(object? sender, ClipboardEventArgs e)
            => ClipboardChanged(e.DataFormat, e.Data);

        public void StartClipboardWatcher()
        {

            if (ClipboardWatcher.isRunning)
                return;

            ClipboardWatcher.Start();
        }    
        
        public void StopClipboardWatcher()
        {
            if (!ClipboardWatcher.isRunning)
                return;

            ClipboardWatcher.Stop();
        }
    }`

Using it as such:
`
        private EventWatcher EventWatcher;

        public void EnableEventWatcher()
        {
            if (EventWatcher == null)
            {
                EventWatcher = new EventWatcher();
                EventWatcher.ClipboardChangedEvent += EventWatcher_ClipboardChangedEvent;
            }

            EventWatcher.StartClipboardWatcher();
        }
        public void DisableEventWatcher()
            => EventWatcher?.StopClipboardWatcher();

        private void EventWatcher_ClipboardChangedEvent(ClipboardContentTypes format, object data)
        {
            Url_TextBox.Text = data.ToString();
            Task.Run(() => StartExecution()).Wait();
        }  `

It pops up an error that i cannot access my Window controls from another thread, but adding a Dispatcher.Invoke does get passed that error.
Only to then completely freeze on the "StartExecution()", no clue why it's bothered about threads ?
Before I was using the NuGet lkst.ClipboardWatcher.WPF, but it only works with a WPF WindowHandle that has been Visible, which won't work for me. That package seems to have no issues with threading, i can subscribe to it's ClipboardChanged event and execute my code just fine.
That put aside, yours seems very clean and has more functions, but I don't get why i can't get it to work.

Any suggestions. (Again, sorry I know it's off topic and doesn't belong here. But couldn't find anohter way to make contact)
Kind regards !

**Comment by javidsho** [comment]
Hi there!

I honestly don't remember this package at all! It must've been something I used privately.

But now I looked it up and it's a port of this library for .NET Core:

https://github.com/justcoding121/windows-user-action-hook

So just follow the examples there and you'd be fine.

**Reply by BourgeoisDirk** [reply]
Thanks for the reply, unfortunately applying the example yielded the exact same result. The event thrown cannot access the Thread that initiated it.
I reverted back to the NuGet package "lkst.ClipboardWatcher.WPF" and made it work by creating a dummy window that is rendered and loaded (thus has a Window Handle) yet is invisible to the user. This one now throws a ClipboardChanged event without the MainWindow ever needing to be loaded and just runs all following code without threading issues.

Thanks again for the quick reply and source material !
